src/backend/replication/logical/message.c

```diff
@@ -93,3 +97,86 @@ logicalmsg_redo(XLogReaderState *record)
 
 	/* This is only interesting for logical decoding, see decode.c. */
 }
+
+/*
+ * NEON: remove AUX object
+ */
+void
+wallog_file_removal(char const* path)
+{
+	char	prefix[MAXPGPATH];
+
+	/* Do not wallog AUX file at replica */
+	if (!XLogInsertAllowed())
+		return;
+
+	snprintf(prefix, sizeof(prefix), "neon-file:%s", path);
+	elog(DEBUG1, "neon: deleting contents of file %s", path);
+
+	/* unlink file */
+	LogLogicalMessage(prefix, NULL, 0, false, true);
+}
+
+/*
+ * NEON: persist file in WAL to save it in persistent storage.
+ */
+void
+wallog_file_descriptor(char const* path, int fd, uint64_t limit)
+{
+	char	prefix[MAXPGPATH];
+	off_t	size;
+	struct stat stat;
+
+	Assert(fd >= 0);
+
+	/* Do not wallog AUX file at replica */
+	if (!XLogInsertAllowed())
+		return;
+
+	if (fstat(fd, &stat))
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not stat file \"%s\": %m", path)));
+	size = stat.st_size;
+
+	elog(DEBUG1, "neon: writing contents of file %s, size %ld", path, (long)size);
+	if (size < 0)
+		elog(ERROR, "Failed to get size of file %s: %m", path);
+
+	if ((uint64_t)size > limit)
+	{
+		elog(WARNING, "Size of file %s %ld is larger than limit %ld", path, (long)size, (long)limit);
+		wallog_file_removal(path);
+	}
+	else
+	{
+		char* buf = palloc((size_t)size);
+		size_t offs = 0;
+		while (offs < size) {
+			ssize_t rc = pread(fd, buf + offs, (size_t)size - offs, offs);
+			if (rc <= 0)
+				elog(ERROR, "Failed to read file %s: %m", path);
+			offs += rc;
+		}
+		snprintf(prefix, sizeof(prefix), "neon-file:%s", path);
+		LogLogicalMessage(prefix, buf, (size_t)size, false, true);
+		pfree(buf);
+	}
+}
+
+void
+wallog_file(char const* path, uint64_t limit)
+{
+	int fd = OpenTransientFile(path, O_RDONLY | PG_BINARY);
+	if (fd < 0)
+	{
+		ereport(LOG,
+				(errcode_for_file_access(),
+				 errmsg("could not open file \"%s\": %m", path)));
```

Comment:
This is highly unexpected. Should probably be ERROR.

Reply:
Changed